### ISSUE TITLE
Feature Request: Ability to wrap lambda handlers

### ISSUE THREAD

**Issue by Kolby-Udacity** [sst/open-next#104]
Hi all, I'm trying to enable Datadog tracing as described here: https://docs.sst.dev/advanced/monitoring

Which requires me to wrap my lambda handlers like:

```ts
import { datadog } from "datadog-lambda-js";

export const handler = datadog(async () => {
  return {
    statusCode: 200,
    body: "Hello World!",
  };
});
```

However, this doesn't work for open-next projects because the lambda handlers are built by open-next.

Based on conversations in Discord the best option appears to be manually wrapping the functions in the build output. This obviously isn't ideal.

I don't know what the API for this feature would look like, but the functionality seems important, at least for my use-case. (If anyone knows of a better solution please let me know).

Maybe a new method in sst.config:

`addDefaultFunctionWrapper: (functionHandler) => datadog(functionHandler)`

**Comment by jayair** [comment]
Oh that's interesting. Let me talk to the team.

**Comment by dreamorosi** [comment]
We are also interested in adding observability to the stack by wrapping the function handlers to apply tracing and contextual logging.

**Comment by juliocorzo** [comment]
Yes, this would be amazing for the for New Relic as well. I tried wrapping the Lambda@Edge functions in serverless-next a long time ago and it wasn't working, but it would be awesome to add that ability, at least for non-edge functions.

**Comment by teriu** [comment]
@Kolby-Udacity @jayair What do you folks think about the API suggested in the OP? I can take a crack at adding this in a PR. We're after this to wrap the handler in the `datadog` HOC too.

**Comment by jayair** [comment]
Oh I forgot to talk to the team, let me get back to you.

**Comment by fwang** [comment]
Reposting from [Discord](https://discord.com/channels/983865673656705025/1027265626085019769/1123361079116976199).

I imagine a file wrapper can be easily added like this:
```bash
npx open-next build \
  && mv .open-next/server-handler/index.mjs .open-next/server-handler/open-next-index.mjs \
  && mv path/to/my/index.mjs .open-next/server-handler/index.mjs
```

And `path/to/my/index.mjs` looks like:
```js
import { handler as openNextHandler } from "./open-next-index.mjs";
export async function handler(event) {
  // do something before
  return await openNextHandler(event);
}
```

Does it need to be built into OpenNext?

/cc @Kolby-Udacity @dreamorosi @juliocorzo @teriu @zds97

**Comment by dreamorosi** [comment]
Thanks for following up on this.

In my use case, if I can do something like the below, it would definitely be an improvement:

```
import { handler as openNextHandler } from "./open-next-index.mjs";

export async function handler(event) {
  // do something before

  try {
    const result = await openNextHandler(event);
    // do something with the result
    return result;
  } catch (err) {
    // do something with the error
    throw err;
  } finally {
    // do something at the very end
  }
}

```

This would allow me to not only do things before but also in case of errors and after. When tracing requests, or creating metrics about them, I need to have access to the result of the function as well as the scope after it has run, otherwise I have no way of determining how long it ran and create a segment/span for it.

Regarding the proposed DX, I suppose it's something and definitely better than nothing.

However (speaking as a library author who'd like to support OpenNext for observability), it would be nice if there was a more ergonomic way of doing it that is abstracted from the final user.

**Comment by Kolby-Udacity** [comment]
> Does it need to be built into OpenNext?

I think we have an opportunity to improve DX and create a more stable solution if we do.

**Comment by Kolby-Udacity** [comment]
Hey @fwang, I gave this a try and started getting 502s and 404s for any external resources my app requested. Any idea why?

Here's what my config looks like:

Note that if I remove `attachDatadog(app);` it works fine, and my custom handler looks like this:

```
import { datadog } from 'datadog-lambda-js';
import { handler as openNextHandler } from './default-handler.mjs';
export async function handler(event) {
  return datadog(await openNextHandler(event));
}
```

```
const sstConfig: SSTConfig = {
  config() {
    return {
      name: 'udacity-web',
      region: 'us-east-1',
    };
  },
  async stacks(app) {
    configureDatadog(app);
    app.stack(NextApp);
    attachRemovalPolicy(app);
    attachDatadog(app);
  },
};

export default sstConfig;

function NextApp({ stack }: StackContext) {
  const { stage } = stack;
  const { parsed: environment } = dotenv.config({ path: envPathMap.get(stage) ?? envPathMap.get('staging') });
  const dataset = process.env.NEXT_PUBLIC_SANITY_DATASET ?? environment?.NEXT_PUBLIC_SANITY_DATASET ?? 'staging';
  const domainName = domainMap.get(stage);

  const serverCachePolicy = new CachePolicy(stack, 'ServerCache', {
    queryStringBehavior: CacheQueryStringBehavior.all(),
    headerBehavior: CacheHeaderBehavior.none(),
    cookieBehavior: CacheCookieBehavior.allowList(JWT_COOKIE),
    defaultTtl: Duration.days(0),
    maxTtl: Duration.days(365),
    minTtl: Duration.days(0),
    enableAcceptEncodingBrotli: true,
    enableAcceptEncodingGzip: true,
  });

  const site = new NextjsSite(stack, 'udacity-web', {
    permissions: ['ssm'],
    edge: false,
    warm: 50,
    waitForInvalidation: true,
    customDomain: {
      domainName: domainName,
      hostedZone: 'web.udacity.com',
    },
    cdk: {
      serverCachePolicy,
    },
    environment: { ...environment, NEXT_PUBLIC_SANITY_DATASET: dataset },
  });

  stack.addOutputs({
    URL: domainName,
  });

  return {
    site,
  };
}

// Ref: https://docs.sst.dev/advanced/removal-policy
function attachRemovalPolicy(app: App) {
  // Remove all resources when non-prod stages are removed
  if (app.stage !== 'prod') {
    app.setDefaultRemovalPolicy('destroy');
  }
}

// Ref: https://docs.sst.dev/advanced/monitoring
function configureDatadog(app: App) {
  app.addDefaultFunctionPermissions([
    new PolicyStatement({
      effect: Effect.ALLOW,
      resources: [process.env.DATADOG_SECRET_ARN ?? ''],
      actions: ['secretsmanager:GetSecretValue'],
    }),
  ]);

  app.setDefaultFunctionProps({
    nodejs: {
      install: ['dd-trace', 'datadog-lambda-js'],
    },
  });
}

async function attachDatadog(app: App) {
  await app.finish();

  // Attach the Datadog construct to each stack
  app.node.children.forEach((stack) => {
    if (stack instanceof Stack) {
      const datadog = new Datadog(stack, 'datadog', {
        // Get the latest version from https://github.com/Datadog/datadog-lambda-js/releases
        nodeLayerVersion: 95,
        // Get the latest version from https://github.com/Datadog/datadog-lambda-extension/releases
        extensionLayerVersion: 45,
        site: 'datadoghq.com',
        apiKeySecretArn: process.env.DATADOG_SECRET_ARN,
        env: app.stage,
        service: app.name,
        version: '1',
        tags: 'team:consumer-eng-team',
      });

      datadog.addLambdaFunctions(stack.getAllFunctions());
    }
  });
}
```

**Comment by jayair** [comment]
I'll ask Frank to follow up.

**Comment by fwang** [comment]
@Kolby-Udacity is the server Lambda getting invoked when u get 502 and 404s?

And if you removed the server function wrapper, while keeping the stacks code as u have above, do u still get 502 and 404s? (I'm curious if the error is caused by stacks code changes or by the function wrapper)

**Comment by Kolby-Udacity** [comment]
> If you removed the server function wrapper, while keeping the stacks code as u have above, do u still get 502 and 404s?

@fwang Yes, it appears just having the `attachDatadog` in my stacks is causing the issue. Removing the wrapped did not fix it.

**Comment by quinnlangille** [comment]
Ran into this issue myself, following the [docs from SST](https://sst.dev/examples/how-to-use-datadog-to-monitor-your-serverless-app.html) it seems that the `stack.getAllFunctions()` method only returns constructs that are an instance of `FN`. It seems that [OpenNext creates CDKFunctions](https://github.com/sst/sst/blob/master/packages/sst/src/constructs/SsrSite.ts#L744). I made an updated version of this method that is working for me:

```typescript
const doGetAllCDKFunctions = (construct: Stack) => {
  const results: Fn[] = [];

  for (const child of construct.node.children) {
    if (child instanceof CDKFN) results.push(child);
    results.push(...doGetAllCDKFunctions(child as Stack));
  }

  return results as LambdaFunction[];
};
```

and I'm using it like this

```
          datadog.addLambdaFunctions(doGetAllCDKFunctions(stack));

```

as a drop in replacement for stack.getAllFunctions. It's working so far, all of the functions created by SST and OpenNext are wrapped by DD as expected. Not sure if it will hold into the future but might be a decent starting place for a first party fix on SST side
